crates/solvers/src/domain/dex/mod.rs

```diff
@@ -101,13 +102,28 @@ impl Swap {
 
     /// Constructs a single order `solution::Solution` for this swap. Returns
     /// `None` if the swap is not valid for the specified order.
-    pub fn into_solution(
+    pub async fn into_solution(
         self,
         order: order::Order,
         gas_price: auction::GasPrice,
         sell_token: Option<auction::Price>,
         score: solution::Score,
+        simulator: &infra::dex::Simulator,
     ) -> Option<solution::Solution> {
+        let gas = if order.class == order::Class::Limit {
+            match simulator.gas(order.owner(), &self).await {
+                Ok(value) => value,
+                Err(err) => {
+                    tracing::warn!(?err, "gas simulation failed");
+                    return None;
+                }
+            }
+        } else {
+            // We are fine with just using heuristic gas for market orders,
+            // since it doesn't really play a role in the final solution.
+            self.gas
+        };
```

Comment:
Sorry I did not review the PR before merging. 

Is it possible to move out gas estimation at the callsite so we can use the accurate `gas estimate` for success_probability calculation also?